minio/minio
# Skip non existent ldap entities while import

## Community Contribution License
All community contributions in this pull request are licensed to the project maintainers
under the terms of the [Apache 2 license](https://www.apache.org/licenses/LICENSE-2.0). 
By creating this pull request I represent that I have the right to license the 
contributions to the project maintainers under the Apache 2 license.

## Description
Dont hard error for non existing ldap entries, rather skip them for import.

## Motivation and Context
We should not hard fail for non existent ldap entities while import and just skip them, so that import succeeds and for existing users/groups things work as expected.

## How to test this PR?
1. Create a MinIO instance and setup LDAP based IDP using [minio-iam-testing](https://github.com/minio/minio-iam-testing)
2. Add couple of service accounts to `uid=dillon,ou=people,ou=swengg,dc=min,dc=io` and `uid=bobfisher,ou=people,ou=hwengg,dc=min,dc=io`. Note they are from groups `hwengg` and `swengg`
3. Export the iam details using `mc admin cluster iam export ALIAS`
4. Edit `ldap/bootstrap.ldif` of `minio-iam-testing` and remove `swengg` relaed entries and build the docker images fresh using `make docker-images`. Deploy new instance of minio-iam-testing using `make docker-run` now
5. Start a fresh MinIO instance and setup LDAP based IDP using the new deployment in step-4
6. Try importing IAM details from step-3 in this MinIO cluster now

MinIO should not show any errors for non existent entities and import should succeed.

It should import the correct ones and you should be able to verify the same using `mc admin user list ALIAS` and `mc admin user svcacct ALIAS USER`. 

Also import command should show the list of skipped entities as below

```
$ mc admin cluster iam import myminio1 EXPORTED-IAM-CFG-ZIP-FILE-NAME
Skipped users: uid=dillon,ou=people,ou=swengg,dc=min,dc=io
Skipped service accounts: dillon-svcacct-2, dillon-svcacct-1
Added policies: readonly, readwrite, writeonly, consoleAdmin, diagnostics
Added service accounts: bobfisher-svcacct-1, bobfisher-svcacct-2
Added policies for sts: uid=bobfisher,ou=people,ou=hwengg,dc=min,dc=io
```
and
```
$ mc admin cluster iam import myminio1 EXPORTED-IAM-CFG-ZIP-FILE-NAME --json
{
 "skipped": {
  "policies": null,
  "users": [
   "uid=dillon,ou=people,ou=swengg,dc=min,dc=io"
  ],
  "serviceAccounts": [
   "dillon-svcacct-1",
   "dillon-svcacct-2"
  ]
 },
 "removed": {
  "policies": null,
  "users": null
 },
 "added": {
  "policies": [
   "readwrite",
   "writeonly",
   "consoleAdmin",
   "diagnostics",
   "readonly"
  ],
  "users": null,
  "serviceAccounts": [
   "bobfisher-svcacct-1",
   "bobfisher-svcacct-2"
  ],
  "stsPolicies": [
   {
    "uid=bobfisher,ou=people,ou=hwengg,dc=min,dc=io": [
     "consoleAdmin"
    ]
   }
  ]
 },
 "failed": {}
}
```

Needs: https://github.com/minio/mc/pull/5032 and https://github.com/minio/madmin-go/pull/302

## Types of changes

## Checklist:
